### PR TITLE
Improve F# support in fsharp branch 

### PR DESCRIPTION
**Pull request by dsyme**
Bringing fsharp branch up-to-date

### CODE REVIEW

**Comment by dsyme** [comment]
CI fails on the ``fsharp`` branch with these 4 errors, all appear to have the same root cause

    https://ci.appveyor.com/project/appsvc/azure-webjobs-sdk-script/build/1.0.14-dxmibwcr/tests

We can still integrate this particular merge though

**Comment by dsyme** [comment]
We have one consistently failing test in CI despite the fact that all the new F# tests are disabled.  This is odd!  

I've checked [the full diff between the azure/dev and dsyme/fsharp branches](https://github.com/Azure/azure-webjobs-sdk-script/compare/dev...dsyme:fsharp) but can't spot anything that would be causing this.  Any help would be appreciated.

**Comment by dsyme** [comment]
OK, I'm now able to start enabling tests and we have some failures specific to F#, e.g.:

```
Microsoft.Azure.WebJobs.Host.FunctionInvocationException : Exception while executing function: Functions.MobileTableOut
---- System.InvalidOperationException : Error while handling parameter item after function returned:
-------- Microsoft.WindowsAzure.MobileServices.MobileServiceInvalidOperationException : Id@ is not a valid identifier. Identifiers must be under 128 characters in length, start with a letter or underscore, and can contain only alpha-numeric and underscore characters.

Microsoft.Azure.WebJobs.Script.Tests.FSharpEndToEndTests.DocumentDB
```

**Comment by mathewc** [comment]
Don, I just pulled your branch, and test ScriptHostManagerTests.RunAndBlock_SetsLastError_WhenExceptionIsThrown passes for me locally. So I don't think you've broken anything - I think we may have a flaky test or timing issue.

Also regarding test coverage: I don't think your FSharp end to end tests need to duplicate all the test scenarios that we have in the C# test suite. Many of those exist to test that the actual bindings work properly, and aren't really language specific. It would suffice for you to just have maybe a few end to end, but you don't need to cover all the different binding types (MobileApp, NotificationHub, etc.).

**Comment by dsyme** [comment]
Thanks @mathewc.  I'll push a bit more on the EndToEnd since we will be adding function templates and docs for all of these in parity with C#.

I'll keep running notes in this PR.  Below is the current list of failures.

```
Microsoft.Azure.WebJobs.Script.Tests.FSharpEndToEndTests.MobileTables
Microsoft.Azure.WebJobs.Script.Tests.FSharpEndToEndTests.ServiceBusQueueTriggerToBlobTest
Microsoft.Azure.WebJobs.Script.Tests.FSharpEndToEndTests.DocumentDB
Microsoft.Azure.WebJobs.Script.Tests.FSharpEndToEndTests.ScriptReference_LoadsScript
Microsoft.Azure.WebJobs.Script.Tests.FSharpEndToEndTests.NotificationHub_Out_Notification
Microsoft.Azure.WebJobs.Script.Tests.FSharpEndToEndTests.MobileTablesTable
Microsoft.Azure.WebJobs.Script.Tests.FSharpEndToEndTests.SharedAssemblyDependenciesAreLoaded
```

**Comment by dsyme** [comment]
Down to three failing tests:

```
Microsoft.Azure.WebJobs.Script.Tests.FSharpEndToEndTests.ServiceBusQueueTriggerToBlobTest
Microsoft.Azure.WebJobs.Script.Tests.dll 1 min 9 sec 992 ms
Microsoft.Azure.WebJobs.Script.Tests.CSharpEndToEndTests.DocumentDB
Microsoft.Azure.WebJobs.Script.Tests.dll 1 min 22 sec 898 ms
Microsoft.Azure.WebJobs.Script.Tests.FSharpEndToEndTests.MobileTables
Microsoft.Azure.WebJobs.Script.Tests.dll 1 min 3 sec 453 ms
```

**Comment by dsyme** [comment]
@fabiocav All three tests still failing give ```System.ApplicationException : Condition not reached within timeout.``` as the only diagnostic.  I assume this means that writing into the test queue/database etc. hasn't triggered the function.  

Let us know if there are better ways to get better diagnostics out of AppVeyor! Thanks :)

**Comment by fabiocav** [comment]
Those are indeed cryptic and hard to diagnose in CI. We're working improve the failure reporting in tests. I'll pull these changes and check into what is happening with those tests.

**Comment by dsyme** [comment]
> Those are indeed cryptic and hard to diagnose in CI. We're working improve the failure reporting in tests. I'll pull these changes and check into what is happening with those tests.

Thanks!  Yes, very cryptic :)  I'm currently [fiddling with things like this](https://github.com/Azure/azure-webjobs-sdk-script/pull/544/commits/47fdeea68e38c03ca6c89a7c0c4164cb0c9ec612) (effectively different ways of writing records in F# - serialization tools can be sensitive to these, e.g. )

File: test/WebJobs.Script.Tests/TestScripts/FSharp/DocumentDBIn/run.fsx [hunk 1]
```diff
@@ -16,9 +16,9 @@ open System
 open Microsoft.Azure.WebJobs.Host
 open Newtonsoft.Json.Linq
 
-[<CLIMutable>]
-type QueueInput =
-    { DocumentId : string }
+//[<CLIMutable>]
+type QueueInput() =
+    member val DocumentId : string = "" with get,set
```
Comment (dsyme) [inline test/WebJobs.Script.Tests/TestScripts/FSharp/DocumentDBIn/run.fsx]: Note to @tomaspetricek and myself: One or both of the changes to DocumentDbIn and DocumentDbOut caused the DocumentDb tests to start passing. So something is sensitive to records v. ``member val``.

Comment (dsyme) [inline test/WebJobs.Script.Tests/TestScripts/FSharp/DocumentDBIn/run.fsx]: Note to @tomaspetricek and myself: One or both of the changes to DocumentDbIn and DocumentDbOut caused the DocumentDb tests to start passing. So something is sensitive to records v. ``member val``.

**Comment by dsyme** [comment]
We are down to two failures - [these changes](https://github.com/Azure/azure-webjobs-sdk-script/pull/544/commits/47fdeea68e38c03ca6c89a7c0c4164cb0c9ec612) caused the DocumentDb test to pass.  I'm not terribly happy that we don't seem to be able to use F# record types though....  And the sensitivity to these sorts of things make me realize that we need to be very careful about specifying and testing the serialization behaviour when interfacing to these data services (and that these kinds of end-to-end tests are very useful when doing that!)

**Comment by dsyme** [comment]
We're now down to one failure (MobileTables).  

We've also been able to do a whole lot of experiments checking which variations of things do/don't work for the various data inputs and outputs, now captured in the specification https://github.com/dsyme/azure-webjobs-sdk-script/blob/dev/docs/fsharp-support.md#functions-triggers-and-serialization-summary

**Comment by dsyme** [comment]
@fabiocav A few things to consider

- We noticed that async methods (with Task returns) don't commute well with the use of out parameters (e.g. see http://stackoverflow.com/questions/18716928/how-to-write-a-async-method-with-out-parameter). Any thoughts on this for C#?

- We removed the need for Out attributes on F# byref parameters - this is the first pace where the F# programming model deviates (allows something extra) over the C# programming model.  This seems OK but we will be sure to write a rationalization for each such case

- It may be sensible to support Async F# methods.  Again this would be an (extra) deviation from the C# model that needs care and would need extra codegen in the dynamically generated thunk to implicitly call Async.StartAsTask.

- Should the tasks have access to a cancellation token?

I'll add these questions to the spec doc https://github.com/dsyme/azure-webjobs-sdk-script/blob/dev/docs/fsharp-support.md#functions-triggers-and-serialization-summary

**Comment by fabiocav** [comment]
@dsyme Here's my first attempt to address your comments/questions above:


> - We noticed that async methods (with Task returns) don't commute well with the use of out parameters (e.g. see http://stackoverflow.com/questions/18716928/how-to-write-a-async-method-with-out-parameter). Any thoughts on this for C#? We could fix this for F# by allowing methods that take "FSharpRef" as parameters though I suspect the underlying azure webjobs SDK still needs a byref out parameter.

To address this, when working with `async`/`Task` returning methods, you can use `IAsyncCollector<T>` or an `ICollector<T>`, which expose `AddAsync` and `Add`, respectively. So instead of something like `out string paramName`, you can use `IAsyncResult<string> paramName` and the SDK will perform the binding with an appropriate collector. F# could use the same pattern/types.

> - We removed the need for Out attributes on F# byref parameters - this is the first pace where the F# programming model deviates (allows something extra) over the C# programming model. This seems OK but we will be sure to write a rationalization for each such case

I agree this would be a good thing, principally if that would feel more familiar to F# developers. Others can weigh in, but in my opinion, the ability to write idiomatic F# would take precedence over parity with what C# exposes.

> - It may be sensible to support Async F# methods. Again this would be an (extra) deviation from the C# model that needs care and would need extra codegen in the dynamically generated thunk to implicitly call Async.StartAsTask.

Are you referring to enhancements to the `FunctionGenerator` class?

> - Should the tasks have access to a cancellation token?

Yes. We automatically pass a cancellation token that is passed into the function if you have a `CancellationToken` argument type. That token will be signaled in cases where the host is shutting down, the function times out (this is work we have in progress), etc.

**Comment by dsyme** [comment]
@fabiocav Thanks for all the answers, we'll update the F# spec with these

**Comment by dsyme** [comment]
OK @fabiocav , I give up on the MobileTablesIn part of the F# test.  I can't see why it's not passing at all - we seem to be doing exactly what C# is doing.  The F# version is failing on this part of the test:

            await WaitForMobileTableRecordAsync("Item", idToCheck, textToCheck);

Could you take a look using local debugging?

For now I'll disable the test for F# so we are green (on a good day - some tests are still flaky)

**Comment by fabiocav** [comment]
I'll spend some time on this today and share what I find.

**Comment by fabiocav** [comment]
@dsyme Unfortunately I didn't get to spend as much time on this as I wanted, but since the Mobile tests for the other languages are working, I'm inclined to say we can work on any remaining cleanup necessary to get this into a state where we're ready to merge to master and address any other F# specific issues there (others also agree).

What we have here is a significant improvement over the current F# implementation and it would be nice to get people playing with this as soon as possible. 

I'm suspecting something is wrong with that test setup. You actually did have a clean run here: https://ci.appveyor.com/project/appsvc/azure-webjobs-sdk-script-y8o14/build/1.0.10348-ukiktsql/tests and the failure was from a different test. So perhaps bringing things back to this state would be good.

Let me know once you feel things are ready to merge and I'll bring this changes in.

File: src/WebJobs.Script/Description/FSharp/FSharpCompiler.cs [hunk 2]
```diff
@@ -50,47 +50,99 @@ public ICompilation GetFunctionCompilation(FunctionMetadata functionMetadata)
...
-            {
-                otherFlags.Add("--debug+");
-                //otherFlags.Add("--tailcalls-");
-            }
+                // Above we have used the C# reference resolver to get the basic set of DLL references for the compilation.
+                //
+                // However F# has its own view on default options. For scripts these should include the
+                // following framework facade references.
+
+                otherFlags.Add("-r:System.Linq.dll"); // System.Linq.Expressions.Expression<T>
```
Comment (dsyme) [inline src/WebJobs.Script/Description/FSharp/FSharpCompiler.cs]: @fabiocav This explicit set of DLL references is unfortunate, but I think OK for now.  We will look at updating FCS some mechanism to give this list

**Comment by dsyme** [comment]
@fabiocav I think this PR is pretty much ready for merging (assuming we get green, or at least no "FSharp" test failures).  We can do the full code review when we merge fsharp --> dev?

**Comment by dsyme** [comment]
@fabiocav OK, the merge commit is green, I think we're good to merge

**Comment by fabiocav** [comment]
Merged into `fsharp`.  🎉 

@mathewc I'll open a PR against `dev` so we can take a look at the changes and iterate there.

**Comment by fabiocav** [comment]
PR #577 has been opened
